gravitee-node-services/gravitee-node-services-upgrader/src/main/java/io/gravitee/node/services/upgrader/UpgraderServiceImpl.java

```diff
@@ -0,0 +1,102 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.node.services.upgrader;
+
+import io.gravitee.common.service.AbstractService;
+import io.gravitee.node.api.Node;
+import io.gravitee.node.api.UpgraderRepository;
+import io.gravitee.node.api.service.UpgraderService;
+import io.gravitee.node.api.upgrader.Upgrader;
+import io.gravitee.node.api.upgrader.UpgraderData;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Kamiel Ahmadpour (kamiel.ahmadpour at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+public class UpgraderServiceImpl
+  extends AbstractService<UpgraderServiceImpl>
+  implements UpgraderService<UpgraderServiceImpl> {
+
+  @Autowired
+  @Lazy
+  private UpgraderRepository upgraderRepository;
+
+  @Value("${upgrade.mode:false}")
+  private boolean upgradeMode;
+
+  private static final Logger logger = LoggerFactory.getLogger(
+    UpgraderServiceImpl.class
+  );
+
+  @Override
+  protected String name() {
+    return "Upgrader service";
+  }
+
+  @Override
+  protected void doStart() throws Exception {
+    super.doStart();
+
+    Map<String, Upgrader> upgraderBeans = applicationContext.getBeansOfType(
+      Upgrader.class
+    );
+    upgraderBeans
+      .values()
+      .stream()
+      .sorted(Comparator.comparing(Upgrader::getOrder))
+      .forEach(
+        upgrader -> {
+          String name = upgrader.getClass().getSimpleName();
+          try {
+            UpgraderData upgraderData = upgraderRepository
+              .findById(upgrader.getClass().getName())
+              .blockingGet();
+            if (upgraderData != null) {
+              logger.info("{} is already applied. it will be ignored.", name);
+            } else {
+              logger.info("Apply {} ...", name);
+              if (upgrader.upgrade()) {
+                upgraderRepository.create(
+                  new UpgraderData(upgrader.getClass().getName(), new Date())
+                );
+              }
+            }
+          } catch (Exception e) {
+            e.printStackTrace();
+            logger.error("Unable to apply {}. Error: ", name, e);
+          }
+        }
+      );
+
+    if (upgradeMode) {
```

Comment:
Why check the `upgradeMode` here ? Could it be more relevant to put it at the beginning instead ?

Reply:
Because even if upgradeMode == null, we still run the first part of the code for backward compatibility

Reply:
Ho! you're right. Even if I remember exactly this rule now, I think this is not really obvious. I suggest to add comment because it sounds a bit odd to have this if here... maybe something like "upgradeAndStop" or "upgradeOnlyMode" that will help to understand this part.